### PR TITLE
Speedup namespace compiler #2 - Kahn's topological sort (relates #721)

### PR DESCRIPTION
**Pull request by Stasik0**
With this reimplementation the generation time goes down from 70 to 30 seconds on nodeset example

### CODE REVIEW

**Comment by Stasik0** [comment]
@markusgraube bump

**Comment by Stasik0** [comment]
there is an arror :(

File: tools/pyUANamespace/ua_namespace.py [hunk 1]
```diff
@@ -527,86 +529,38 @@ def printDotGraphWalk(self, depth=1, filename="out.dot", rootNode=None, followIn
...
+      u = Q.pop()          # choose node of zero in-degree
+      L.append(u)          # and 'remove' it from graph
+      for ref in u.getReferences():
+       if isinstance(ref.target(), opcua_node_t):
+         if ref.referenceType().browseName() in relevant_references and ref.isForward():
+           in_degree[ref.target()] -= 1
+           if in_degree[ref.target()] == 0:
+             Q.appendleft(ref.target())
+    if len(L) == len(self.nodes):
+        self.node = L
```
Comment (markusgraube) [inline tools/pyUANamespace/ua_namespace.py]: Shouldn't this be `self.nodes = L`?

File: tools/pyUANamespace/ua_namespace.py [hunk 2]
```diff
@@ -527,86 +529,38 @@ def printDotGraphWalk(self, depth=1, filename="out.dot", rootNode=None, followIn
...
+        Q.appendleft(u)
+ 
+    L = []     # list for order of nodes
+    
+    while Q:
+      u = Q.pop()          # choose node of zero in-degree
+      L.append(u)          # and 'remove' it from graph
+      for ref in u.getReferences():
+       if isinstance(ref.target(), opcua_node_t):
+         if ref.referenceType().browseName() in relevant_references and ref.isForward():
```
Comment (markusgraube) [inline tools/pyUANamespace/ua_namespace.py]: The previous code took also the subtypes of the relevant references into account. Maybe, we should add this here too

**Comment by Stasik0** [comment]
@markusgraube note that I had to add HasComponent to relevant references, since othervise the order was broken (components were initialized before their parents)

**Comment by markusgraube** [comment]
@Stasik0 Really nice


Maybe, we could add all subtypes of HierarchicalReferences to relevant_references if this makes the sorting not too slow
```
self.getSubTypesOf(self.getNodeByBrowseName("HierarchicalReferences")
```
At least this shoul be still a DAG

**Comment by Stasik0** [comment]
I have added subtypes now, however the getSubTypesOf function crashes if i give HasComponent to it (i had to woraround this) the same is for self.getSubTypesOf(self.getNodeByBrowseName("HierarchicalReferences")

but the generator works on my example and on NS0, how bout you?

> At least this shoul be still a DAG
i am not sure ) HasComponent and HasComponentOf are both hierarchical, but the graph is cyclic...

**Comment by Stasik0** [comment]
unfortunately, "subtype_refrences" is also empty on my examples :(

**Comment by Stasik0** [comment]
I got some other small optimizations, compile time is down to 17 seconds, @markusgraube could we also bake your improvements in? then we will get nscompiler on speed

**Comment by jpfr** [comment]
I got some other small optimizations, compile time is down from 70 to 19 seconds.
:+1: Really nice!

**Comment by markusgraube** [comment]
@Stasik0 

The sorting doesn't work for ns0
> Node graph is circular on the specified references

Thus, I can't push the sorting after the ignoring phase

**Comment by jpfr** [comment]
> The sorting doesn't work for ns0

There are some well-known circular dependencies in ns0.
Since we check for circular dependencies during AddNodes, we need some "bootstrapping" in the server.
E.g. here: https://github.com/open62541/open62541/blob/0.2/src/server/ua_server.c#L762

We somehow need to whitelist those nodes. Imho @ichrispa already did that in his current sort mechanism.

By the way, if we could assume a minimal set of "boostrapped" nodes in the server, the namespace generation could use the public API only. That would make things a lot easier to handle for end-users bringing their own namespace definitions.

**Comment by Stasik0** [comment]
@jpfr this bootstrapping fails somehow (shadowed definition etc)
@markusgraube I reproduced some @ichrispa's code I dont fully understand, but now it seems to work, could you try?

**Comment by Stasik0** [comment]
btw ns0 generation seems a bit broken in 0.2 and here -> look at duplicate objects under "Server"

**Comment by markusgraube** [comment]
@Stasik0 Yes, ns0 is compiling now. However, this solution is not great because the topological sorting just sorts 76 elements of 1516, which are exactly these elements which have no dependencies at all.
Then, we could just omit the sorting at all.

By the way: How much slower is bootstrapping node generation than using the high-level API? Or is it a matter of code and memory size that we want to avoid bootstrapping?

**Comment by jpfr** [comment]
Bootstrapping should be faster than going by the AddNodes service. By adding nodes directly to the nodestore with no consistency checks.
But access to the nodestore is an internal API that is not intended for use by the end users.

**Comment by Stasik0** [comment]
@markusgraube well, the behaviour is exactly as it was with the current implementation, probably we need to return to the specification of HasSubtype, HasTypeDefinitoin, Organizes, HasComponent and some more to get it better... However in this case the bootstrapping breaks as you have seen and I have no clue how to or time to fix it ATM.

Removing sorting completely gives a gain of about 2 sec with server_nodeset.xml. For complete ns0 the gain is not even measurable (i suspect it is also about 2 sec). I'd let it in and consider restricting relations and reworking bootstrapping at a later point of time.

**Comment by Stasik0** [comment]
i give up atm, the bootstrapping is broken and code is too complex

**Comment by Stasik0** [comment]
okay: summary of bugs 

(1)
bootstrapping has to be fixed at first.

if you comment out this line
https://github.com/open62541/open62541/pull/725/commits/2cb12485e8e1d1f98781926e53a26b42ab3e6a06#diff-81712a77990bcdaef01316de80879f87R555

the compilation of ns0 will break, however it should not be the case since NS-compiler has to cope with any order

(2) self.getSubTypesOf(currentNode=tn) return empty lists

I vote for letting the "sort" (which is actually not sorting) inside, work on these bugs and make a proper sorting later, by removing the marked line

 @open62541/open62541-core okay?

**Comment by jpfr** [comment]
> I vote for letting the "sort" (which is actually not sorting) inside, work on these bugs and make a proper sorting later, by removing the marked line

When is later? Post 0.2?

**Comment by Stasik0** [comment]
okay, guys. After the last spin, I made a proper topological sort, it works at least for server_nodeset.xml pretty good. Even the generation of NS0 seems to work as before.

However, I could not add a fix by @markusgraube, somehow sorting only not-ignored nodes breaks the order (just try by using a re-sorted server_nodeset.xml)

for nodeset_example the time is down to 15s

okay to merge? Please give it a try with some reordered XML files

**Comment by Stasik0** [comment]
@ichrispa :+1: or :-1: ?

**Comment by markusgraube** [comment]
@Stasik0 #what does this PR for the [minimal example model from #729](https://gist.github.com/markusgraube/058f0535100e7bf9c9ea448b0253289f)? It has a circular dependency, but should be able to create with high level api

**Comment by ichrispa** [comment]
As with #729: Please don't replace djikstra, just add an alternative. A commandline switch to pick the sorting algorithm would be nice.

**Comment by Pro** [comment]
@ichrispa @markusgraube @Stasik0 any update on this?
Would be really nice if we can integrate all the Namespace Compiler PRs or close them if they are not relevant anymore.

**Comment by Pro** [comment]
Should we just close this in favour of #739 ?

**Comment by Stasik0** [comment]
My parts seem to be in 739, so +1 to close

**Pull request closed without merging**